### PR TITLE
Deprecate torch.triangular_solve

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63570

There is a use of `at::triangular_solve_out` in the file
`torch/csrc/jit/tensorexpr/external_functions.cpp` that I have not dared
to move to `at::linalg_solve_triangular_out`.

**Deprecation note:**

This PR deprecates the `torch.triangular_solve` function in favor of
`torch.linalg.solve_triangular`. An upgrade guide is added to the
documentation for `torch.triangular_solve`.

Note that it DOES NOT remove `torch.triangular_solve`, but
`torch.triangular_solve` will be removed in a future PyTorch release.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

Differential Revision: [D32618035](https://our.internmc.facebook.com/intern/diff/D32618035)

### CODE REVIEW

**Review by IvanYashchuk — APPROVED** [review]

**Review by nikitaved — APPROVED** [review]

**Review by anjali411 — APPROVED** [review]

File: test/backward_compatibility/check_backward_compatibility.py [hunk 1]
```diff
@@ -48,7 +48,6 @@
     ("aten::slice_backward", datetime.date(9999, 1, 1)),
     ("aten::diagonal_backward", datetime.date(9999, 1, 1)),
     ("aten::rowwise_prune", datetime.date(9999, 1, 1)),
-    ("aten::_triangular_solve_helper", datetime.date(9999, 1, 1)),
```
Comment (mruberry) [inline test/backward_compatibility/check_backward_compatibility.py:51]: Yikes, so much for this file being meaningful.

**Comment by mruberry** [comment]
cc @Balandat -- this will require updates if you don't want to throw a warning
cc @ZolotukhinM for the jit usage of triangular solve

**Review by mruberry — APPROVED** [review]

**Comment by ZolotukhinM** [comment]
cc @Chillee, I think you've used it somewhere?

**Comment by anjali411** [comment]
`test_triangular_solve_out_errors_and_warnings_cpu_float64` fails because we are throwing an extra warning

**Comment by Lezcano** [comment]
@anjali411 fixed. I also updated a few new used that we had in the codebase of the older `triangular_solve`.

**Comment by anjali411** [comment]
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32618035).

**Pull request closed without merging**